Detectors/ZDC/simulation/src/Digits2Raw.cxx

```diff
@@ -0,0 +1,874 @@
+    unsigned s10 : 12;
+    unsigned s11 : 12;
+    unsigned empty_4 : 16;
+    UInt_t empty_5;
+  };
+
+  ULong64_t lsb = val;
+  ULong64_t msb = val >> 64;
+  UInt_t a = word[0];
+  UInt_t b = word[1];
+  UInt_t c = word[2];
+  //UInt_t d=(msb>>32)&0xffffffff;
+  //printf("\n%llx %llx ",lsb,msb);
+  //printf("\n%8x %8x %8x %8x ",d,c,b,a);
+  if ((a & 0x3) == 0) {
+    union {
+      unsigned __int128 w;
+      struct zdc_payload_v0_w0 f;
+    } w0;
+    UInt_t myorbit = (val >> 48) & 0xffffffff;
+    UInt_t mybc = (val >> 36) & 0xfff;
+    if (myorbit != last_orbit || mybc != last_bc) {
+      printf("Orbit %9u bc %4u\n", myorbit, mybc);
+      last_orbit = myorbit;
+      last_bc = mybc;
+    }
+    printf("%04x %08x %08x ", c, b, a);
+    UInt_t hits = (val >> 24) & 0xfff;
+    Int_t offset = (lsb >> 8) & 0xffff - 32768;
+    Float_t foffset = offset / 8.;
+    UInt_t board = (lsb >> 2) & 0xf;
+    UInt_t ch = (lsb >> 6) & 0x3;
+    //printf("orbit %9u bc %4u hits %4u offset %+6i Board %2u Ch %1u", myorbit, mybc, hits, offset, board, ch);
+    printf("orbit %9u bc %4u hits %4u offset %+8.3f Board %2u Ch %1u", myorbit, mybc, hits, foffset, board, ch);
+    if (board >= NModules)
+      printf(" ERROR with board");
+    if (ch >= NChPerModule)
+      printf(" ERROR with ch");
+    if (moduleConfig) {
+      auto id = moduleConfig->modules[board].channelID[ch];
+      if (id >= 0 && id < NChannels)
+        printf(" %s", ChannelNames[id].data());
+      else
+        printf(" error with ch id");
+    }
+  } else if ((a & 0x3) == 1) {
+    printf("%04x %08x %08x ", c, b, a);
+    union {
+      unsigned __int128 w;
+      struct zdc_payload_v0_w1 f;
+    } w1;
+    printf("     %s %s %s %s ", a & 0x10 ? "A0" : "  ", a & 0x20 ? "A1" : "  ", a & 0x40 ? "A2" : "  ", a & 0x80 ? "A3" : "  ");
+    printf("0-5 ");
+    Short_t s[6];
+    val = val >> 8;
+    for (Int_t i = 0; i < 6; i++) {
+      s[i] = val & 0xfff;
+      if (s[i] > 2047)
+        s[i] = s[i] - 4096;
+      val = val >> 12;
+    }
+    printf(" %5d %5d %5d %5d %5d %5d", s[0], s[1], s[2], s[3], s[4], s[5]);
+  } else if ((a & 0x3) == 2) {
+    printf("%04x %08x %08x ", c, b, a);
+    union {
+      unsigned __int128 w;
+      struct zdc_payload_v0_w2 f;
+    } w2;
+    printf("%s %s %s %s %s %s ", a & 0x4 ? "H" : " ", a & 0x8 ? "TM" : "  ", a & 0x10 ? "T0" : "  ", a & 0x20 ? "T1" : "  ", a & 0x40 ? "T2" : "  ", a & 0x80 ? "T3" : "  ");
+    printf("6-b ");
+    Short_t s[6];
+    val = val >> 8;
+    for (Int_t i = 0; i < 6; i++) {
+      s[i] = val & 0xfff;
+      if (s[i] > 2047)
+        s[i] = s[i] - 4096;
+      val = val >> 12;
+    }
+    printf(" %5d %5d %5d %5d %5d %5d", s[0], s[1], s[2], s[3], s[4], s[5]);
+  } else if ((a & 0x3) == 3) {
+    printf("%04x %08x %08x ", c, b, a);
+    printf("HB ");
+  }
+  printf("\n");
+}
+
+//______________________________________________________________________________
+void Digits2Raw::emptyBunches(std::bitset<3564>& bunchPattern)
```

Comment:
Is this supposed to provide empty bunches for simulated data? Note that the simulated bunch filling is stored and can be used as 
````
const auto* digctx = o2::steer::DigitizationContext::loadFromFile("collisioncontext.root");
const auto& bcfill = digctx->getBunchFilling();
for (int i = 0; i<o2::constants::lhc::LHCMaxBunches; i++) {
    if (!bf.testBC(i)) {
       printf("bc %d is not filled\n",i);
    }
````

Reply:
I'm not sure of the most appropriate position when to load the filling scheme.
At present the filling scheme is extracted using this code is in the workflow executabile digi2raw.cxx
This function restricts the definition of empty bunch to remove bunches that are close to a B bunch or with beam gas from back of the calorimeter. In the future one should remove also A and C bunches (at least for the actual data taking).